### PR TITLE
Introduce modules.ValidateResource for Cloud-specific validation

### PR DESCRIPTION
**Pull request by andrejtokarcik**
Following the suggestion of https://github.com/gravitational/teleport/pull/6472#discussion_r625239022.

Unfortunately, there seems to be no common point allowing to run the validation on all `Resource`s in general. Hence the call to `ValidateResource` is added just to the individual setters of the configuration resources.

### CODE REVIEW

File: lib/modules/modules.go [hunk 1]
```diff
@@ -74,6 +75,8 @@ type Modules interface {
 	Features() Features
 	// BuildType returns build type (OSS or Enterprise)
 	BuildType() string
+	// ValidateResource performs additional resource checks.
+	ValidateResource(types.Resource) error
```
Comment (awly) [inline lib/modules/modules.go:79]: WDYT about making this a top-level function?
Something like:
```go
func ValidateResource(r types.Resource) error {
  if !GetModules().Features().Cloud {
    return nil
  }
  // validation goes here...
}
```

That way we don't need to separately update `teleport.e`

**Review by Joerger — APPROVED** [review]

**Review by alex-kovoy — APPROVED** [review]

File: lib/modules/modules.go (hunk 1 above)
Comment (andrejtokarcik) [reply lib/modules/modules.go:79]: To satisfy the current requirements a top-level function would suffice. I'm not sure it completely eliminates the need for `teleport.e` updates though -- it depends on how "tight" we want our tests to be. If the resource checks should be tested with `enterpriseModules` directly no matter what, the tests will still have to go to `teleport.e`.

Comment (awly) [reply lib/modules/modules.go:79]: You can populate the `Cloud` flag with:
```
type cloudModules struct {
  *defaultModules
}

func (m cloudModules) Features() Features {
  f := m.defaultModules.Features()
  f.Cloud = true
  return f
}
```

Comment (andrejtokarcik) [reply lib/modules/modules.go]: @awly Converted `ValidateResource` into a top-level function. I had to do additional adjustments to `modules_test` due to cyclic imports, let me know if it's acceptable.

**Review by awly — APPROVED** [review]

**Review by r0mant — APPROVED** [review]
🤖

**Pull request merged**